bridge-design.md

```diff
@@ -0,0 +1,167 @@
+ArrayBuffers.
+
+There will be no direct calls from javascript to rust code, and little binding
+code in general, although we do need a handful of for purposes of managing
+buffers, and to allow the VM thread to go to sleep when it has no events to
+process for a prolonged period of time.
+
+## Inter-thread communication
+
+The VM thread sends messages to the backend thread by means of a
+SharedArrayBuffer to which both threads have access. Messages are written
+linearly into this buffer (each prefixed by the message length). The backend
+thread sends the messages back the same way.
+
+The very last int32_t slot of the buffer (let's call it `writer_pos`) is used
+for synchronization. Usually it indicates how many bytes have been written by
+the producer at the beginning of the buffer, which is also the offset at which
+the sender will write the next message.
+
+The consumer spins, polling this field to detect messages appearing in the
+buffer. If no new messages appear after spinning for some time, the consumer
+thread goes to sleep, and has to be awoken by the producer thread by some other
+mechanism (e.g. a futex).
+
+When the buffer is (almost) full, the producer allocates a new buffer and places
+a special message in the buffer to point the consumer to the next buffer.
+
+When the consumer has processed all messages in the buffer, and the producer has
+moved on to the next buffer, the buffer is released (what this means exactly
+depends on whether the consumer is the VM thread or the backend thread - see
+below).
+
+## Message serialization
+
+The messages placed in the ring buffer may be of variable length, and are
+encoded in some binary format. The exact message protocol is out of scope for
+this document.
+
+## Queue consistency
+
+Failure to send a message should not leave the message queue in an inconsistent
+state, nor should it evoke resource or memory leaks. Some failure modes: *
+Exception thrown during serialization. * Receiving end does not have sufficient
+memory available to process the message. * Message is not understood by the
+receiving end.
+
+## Transferring buffers
+
+It wouldn't be optimal to place all data in the ring buffer, e.g. when sending a
+large chunk of data, the overhead of copying it into and out of the command
+buffer might be significant. When data is already in a buffer, we may want to
+simply reference them, rather than copying their contents. Likewise, when data
+is read, ideally it'd be written directly into a buffer that can be passed off
+to the user, rather than in the command buffer.
+
+Therefore an inter-thread message needs to be able to reference a buffer. Since
+messages are written into a bytearray, this reference must be a number; an
+object reference is of no use because it can't be written into an ArrayBuffer.
+The backend thread needs to be able to map this reference to a pointer and a
+length (without calling V8 APIs).

```

Comment:
So if I understand this correctly it will write an index into some table - which then has pointers to the actual data? The table is in the privlaged section?